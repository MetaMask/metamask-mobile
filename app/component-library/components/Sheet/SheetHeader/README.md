# SheetHeader

SheetHeader is a header component and is currently used within the [BottomSheet](../BottomSheet/BottomSheet.tsx) component.

## Props

### `title`

Sheet title.

| <span style="color:gray;font-size:14px">TYPE</span> | <span style="color:gray;font-size:14px">REQUIRED</span> |
| :-------------------------------------------------- | :------------------------------------------------------ |
| string                                              | Yes                                                     |

### `onBack`

Optional callback when back button is pressed. The back button appears when this property is set.

| <span style="color:gray;font-size:14px">TYPE</span> | <span style="color:gray;font-size:14px">REQUIRED</span> |
| :-------------------------------------------------- | :------------------------------------------------------ |
| function                                            | No                                                      |

### `actionButtonOptions`

Optional action option, which includes a callback when the action button is pressed. The action button appears when this property is set.

| <span style="color:gray;font-size:14px">TYPE</span>         | <span style="color:gray;font-size:14px">REQUIRED</span> |
| :---------------------------------------------------------- | :------------------------------------------------------ |
| [SheetHeaderActionButtonOptions](./SheetHeader.types.ts#L1) | No                                                      |

## Usage

```javascript
// Replace with relative import.
import SheetHeader from 'app/component-library/components/SheetHeader';

<SheetHeader
  onBack={ONBACK_HANDLER}
  actionButtonOptions={{ label: ACTION_LABEL, onPress: ONPRESS_HANDLER }}
  title={TITLE_LABEL}
/>;
```
