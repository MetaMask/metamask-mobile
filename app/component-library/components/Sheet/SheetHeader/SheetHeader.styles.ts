// Third party dependencies.
import { StyleSheet } from 'react-native';

// External dependencies.
import { Theme } from '../../../../util/theme/models';

/**
 * Style sheet function for SheetHeader component.
 *
 * @param params Style sheet params.
 * @param params.theme App theme from ThemeContext.
 * @param params.vars Inputs that the style sheet depends on.
 * @returns StyleSheet object.
 */
const styleSheet = (params: { theme: Theme }) => {
  const { theme } = params;
  const { colors } = theme;
  return StyleSheet.create({
    base: {
      flexDirection: 'row',
      alignItems: 'center',
      justifyContent: 'space-between',
      margin: 16,
      backgroundColor: colors.background.default,
      height: 32,
    },
    leftAccessory: {
      flex: 1,
    },
    rightAccessory: {
      flex: 1,
      alignItems: 'flex-end',
    },
  });
};

export default styleSheet;
