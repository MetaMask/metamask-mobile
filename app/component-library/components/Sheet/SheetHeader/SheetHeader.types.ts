/**
 * Options for displaying the action button on the right of the title.
 */
export interface SheetHeaderActionButtonOptions {
  label: string;
  onPress: () => void;
}

/**
 * SheetHeader component props.
 */
export interface SheetHeaderProps {
  /**
   * Sheet title.
   */
  title: string;
  /**
   * Optional callback when back button is pressed. The back button appears when this property is set.
   */
  onBack?: () => void;
  /**
   * Optional action options, which includes a callback when the action button is pressed. The action button appears when this property is set.
   */
  actionButtonOptions?: SheetHeaderActionButtonOptions;
}
