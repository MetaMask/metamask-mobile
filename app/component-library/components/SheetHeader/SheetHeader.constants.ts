// Test IDs.
export const SHEET_HEADER_BACK_BUTTON_ID = 'sheet-header-back-button';
export const SHEET_HEADER_ACTION_BUTTON_ID = 'sheet-header-action-button';
