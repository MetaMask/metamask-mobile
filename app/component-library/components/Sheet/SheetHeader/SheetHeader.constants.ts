/* eslint-disable no-console */
// Internal dependencies.
import { SheetHeaderProps } from './SheetHeader.types';

// Test IDs.
export const SHEET_HEADER_BACK_BUTTON_ID = 'sheet-header-back-button';
export const SHEET_HEADER_ACTION_BUTTON_ID = 'sheet-header-action-button';

// Sample consts
export const SAMPLE_SHEETHEADER_PROPS: SheetHeaderProps = {
  title: 'Sample Sheet Title',
  onBack: () => console.log('onBack clicked'),
  actionButtonOptions: {
    label: 'button Label',
    onPress: () => console.log('label clicked'),
  },
};
