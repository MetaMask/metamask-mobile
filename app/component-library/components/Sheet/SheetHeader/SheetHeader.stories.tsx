// Third party dependencies.
import React from 'react';
import { storiesOf } from '@storybook/react-native';
import { boolean, text } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';

// Internal dependencies.
import SheetHeader from './SheetHeader';

storiesOf('Component Library / SheetHeader', module).add('Default', () => {
  const groupId = 'Props';
  const includesOnBack = boolean('Includes onBack', false, groupId);
  const includesActionButtonOptions = boolean(
    'Includes actionButtonOptions',
    false,
    groupId,
  );
  const onBack = includesOnBack ? action('onPress') : undefined;
  const onPress = action('onPress');
  const actionButtonLabel = includesActionButtonOptions
    ? text('actionButtonOptions.label', 'Action', groupId)
    : '';
  const titleLabel = text('title', 'Title', groupId);
  const actionButtonOptions = includesActionButtonOptions
    ? {
        label: actionButtonLabel,
        onPress,
      }
    : undefined;

  return (
    <SheetHeader
      onBack={onBack}
      actionButtonOptions={actionButtonOptions}
      title={titleLabel}
    />
  );
});
