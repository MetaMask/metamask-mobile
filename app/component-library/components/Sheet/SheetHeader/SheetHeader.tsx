/* eslint-disable react/prop-types */

// Third party dependencies.
import React from 'react';
import { View } from 'react-native';

// External dependencies.
import Text, { TextVariant } from '../../Texts/Text';
import Button, { ButtonVariants } from '../../Buttons/Button';
import ButtonIcon, { ButtonIconVariants } from '../../Buttons/ButtonIcon';
import { IconName } from '../../Icon';
import { useStyles } from '../../../hooks';

// Internal dependencies.
import styleSheet from './SheetHeader.styles';
import { SheetHeaderProps } from './SheetHeader.types';
import {
  SHEET_HEADER_ACTION_BUTTON_ID,
  SHEET_HEADER_BACK_BUTTON_ID,
} from './SheetHeader.constants';

const SheetHeader = ({
  title,
  onBack,
  actionButtonOptions,
  ...props
}: SheetHeaderProps) => {
  const { styles } = useStyles(styleSheet, {});

  return (
    <View style={styles.base} {...props}>
      <View style={styles.leftAccessory}>
        {onBack && (
          <ButtonIcon
            testID={SHEET_HEADER_BACK_BUTTON_ID}
            variant={ButtonIconVariants.Secondary}
            onPress={onBack}
            iconName={IconName.ArrowLeftOutline}
          />
        )}
      </View>
      <Text variant={TextVariant.HeadingMD}>{title}</Text>
      <View style={styles.rightAccessory}>
        {actionButtonOptions && (
          <Button
            variant={ButtonVariants.Link}
            testID={SHEET_HEADER_ACTION_BUTTON_ID}
            onPress={actionButtonOptions.onPress}
            textVariant={TextVariant.BodyMD}
            label={actionButtonOptions.label}
          />
        )}
      </View>
    </View>
  );
};

export default SheetHeader;
