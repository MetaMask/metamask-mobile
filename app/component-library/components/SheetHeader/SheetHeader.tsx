/* eslint-disable react/prop-types */

// Third party dependencies.
import React from 'react';
import { View } from 'react-native';

// External dependencies.
import Link from '../Link';
import BaseText, { BaseTextVariant } from '../BaseText';
import IconButton, { IconButtonVariant } from '../IconButton';
import { IconName } from '../Icon';
import { useStyles } from '../../hooks';

// Internal dependencies.
import styleSheet from './SheetHeader.styles';
import { SheetHeaderProps } from './SheetHeader.types';
import {
  SHEET_HEADER_ACTION_BUTTON_ID,
  SHEET_HEADER_BACK_BUTTON_ID,
} from './SheetHeader.constants';

const SheetHeader = ({
  title,
  onBack,
  actionButtonOptions,
  ...props
}: SheetHeaderProps) => {
  const { styles } = useStyles(styleSheet, {});

  return (
    <View style={styles.base} {...props}>
      <View style={styles.leftAccessory}>
        {onBack && (
          <IconButton
            testID={SHEET_HEADER_BACK_BUTTON_ID}
            variant={IconButtonVariant.Secondary}
            onPress={onBack}
            icon={IconName.ArrowLeftOutline}
          />
        )}
      </View>
      <BaseText variant={BaseTextVariant.sHeadingMD}>{title}</BaseText>
      <View style={styles.rightAccessory}>
        {actionButtonOptions && (
          // TODO - Replace Link with ButtonTertiary once new variant is ready/
          <Link
            testID={SHEET_HEADER_ACTION_BUTTON_ID}
            onPress={actionButtonOptions.onPress}
            variant={BaseTextVariant.sBodyMD}
          >
            {actionButtonOptions.label}
          </Link>
        )}
      </View>
    </View>
  );
};

export default SheetHeader;
