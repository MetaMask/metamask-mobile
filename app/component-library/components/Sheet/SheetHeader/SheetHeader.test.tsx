// Third party dependencies.
import React from 'react';
import { shallow } from 'enzyme';

// Internal dependencies.
import SheetHeader from './SheetHeader';
import {
  SHEET_HEADER_ACTION_BUTTON_ID,
  SHEET_HEADER_BACK_BUTTON_ID,
} from './SheetHeader.constants';

describe('SheetHeader', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<SheetHeader title={'Title'} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render back button', () => {
    const wrapper = shallow(<SheetHeader onBack={jest.fn} title={'Title'} />);
    const backButton = wrapper.findWhere(
      (node) => node.prop('testID') === SHEET_HEADER_BACK_BUTTON_ID,
    );
    expect(backButton.exists()).toBe(true);
  });

  it('should render action button', () => {
    const wrapper = shallow(
      <SheetHeader
        title={'Title'}
        actionButtonOptions={{ label: 'Action', onPress: jest.fn }}
      />,
    );
    const actionButton = wrapper.findWhere(
      (node) => node.prop('testID') === SHEET_HEADER_ACTION_BUTTON_ID,
    );
    expect(actionButton.exists()).toBe(true);
  });
});
