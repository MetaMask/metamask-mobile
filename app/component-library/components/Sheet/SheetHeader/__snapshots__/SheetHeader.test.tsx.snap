// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SheetHeader should render correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flexDirection": "row",
      "height": 32,
      "justifyContent": "space-between",
      "margin": 16,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  />
  <Text
    variant="sHeadingMD"
  >
    Title
  </Text>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flex": 1,
      }
    }
  />
</View>
`;
